### PR TITLE
Add action contracts for migrating live orbit chains to espresso

### PR DESCRIPTION
**Pull request by zacshowa**
### This PR:

This PR adds action upgrade contracts to the orbit-action repo to upgrade orbit chains to be compatible with the espresso sequencing marketplace.

### This PR does not:

Include scripts to deploy and perform these action upgrades. Set the chain config in the arbOS upgrade.

### Key places to review:
EspressoOspMigrationAction.sol and EspressoMigrationAction.t.sol.

forge test in the repo's top level

EspressoOspMigrationAction.sol

### CODE REVIEW

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    constructor(uint64 _newArbOSVersion, uint64 _upgradeTimestamp) {
+        newArbOSVersion = _newArbOSVersion;
+        upgradeTimestamp = _upgradeTimestamp;
+    }
+
+    function perform() external {
+        ArbOwner arbOwner = ArbOwner(0x0000000000000000000000000000000000000070);
+        arbOwner.scheduleArbOSUpgrade({
+            newVersion: newArbOSVersion,
```
Comment (Sneh1999) [inline contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: Shouldnt the new version and timestamp be function arguments instead of constructor arguments? Then we can use the same script for multiple upgrade IMO

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+import "forge-std/Script.sol";
+import "@arbitrum/nitro-contracts/src/challenge/IChallengeManager.sol";
+import "@arbitrum/nitro-contracts/src/rollup/IRollupCore.sol";
+import "@arbitrum/nitro-contracts/src/rollup/IRollupAdmin.sol";
+
+import "@openzeppelin/contracts/proxy/transparent/ProxyAdmin.sol";
+import "@openzeppelin/contracts/utils/Address.sol";
+
+error IncorrectWasmModuleRoot(bytes32 incorrectAddr);
```
Comment (Sneh1999) [inline contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: The errors should be a part of the contract, so I think you should add them inside the contract object

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        //Handle assertions in the perform functoin as we shouldn't be storing local state for delegated calls.
+        if(newWasmModuleRoot == bytes32(0)){
+            revert IncorrectWasmModuleRoot(newWasmModuleRoot);
+        }
+
+        if(currentWasmModuleRoot == bytes32(0)){
+            revert IncorrectWasmModuleRoot(currentWasmModuleRoot);
+        }
+
+        if(Address.isContract(newOspEntry) == false){
```
Comment (Sneh1999) [inline contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: nit: can we do ` if(!Address.isContract(newOspEntry))`

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            revert AddressIsNotContract(newOspEntry);
+        }
+
+        if(Address.isContract(currentOspEntry) == false){
```
Comment (Sneh1999) [inline contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: same here

File: contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export NEW_OSP_ENTRY="0x7B5F0B437EE68A22992DdD629AA22525Fc5dfa9A"
```
Comment (Sneh1999) [inline contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: Nit: I dont think you need `export` in front of the environment variables ( example [here](https://github.com/EspressoSystems/nitro-contracts/blob/integration/.env.sample.goerli#L1))

File: contracts/parent-chain/contract-upgrades/example-env/.env-schedule-arbos-upgrade.example [hunk 6]
```diff
@@ -0,0 +1 @@
+export UPGRADE_TIMESTAMP = "1723664126"
```
Comment (Sneh1999) [inline contracts/parent-chain/contract-upgrades/example-env/.env-schedule-arbos-upgrade.example:1]: same here

File: contracts/tests/EspressoMigrationAction.t.sol [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            bridge: new ERC20Bridge(),
+            sequencerInbox: new SequencerInbox(MAX_DATA_SIZE, dummyReader4844, true),
+            inbox: new ERC20Inbox(MAX_DATA_SIZE),
+            rollupEventInbox: new ERC20RollupEventInbox(),
+            outbox: new ERC20Outbox()
+        });
+
+
+    /* solhint-disable func-name-mixedcase */
+    //create items needed for a rollup and deploy it. This code is lovingly borrowed from the rollupcreator.t.sol foundry test.
```
Comment (Sneh1999) [inline contracts/tests/EspressoMigrationAction.t.sol:81]: nice!

**Comment by alysiahuggins** [comment]
Hey @zacshowa, can you also add a readme that explains these code additions and how to run the scripts

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            timestamp: upgradeTimestamp
+        });
+    }
+}
+
+contract EspressoArbOSUpgrade is UpgradeArbOSVersionAtTimestampAction, Script   {
+    constructor()
+        UpgradeArbOSVersionAtTimestampAction(
+            35,
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: when is this called? curious about this hardcoded arbOsVersion

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    uint64 public immutable newArbOSVersion;
+    uint64 public immutable upgradeTimestamp;
+
+    constructor(uint64 _newArbOSVersion, uint64 _upgradeTimestamp) {
+        newArbOSVersion = _newArbOSVersion;
+        upgradeTimestamp = _upgradeTimestamp;
+    }
+
+    function perform() external {
+        ArbOwner arbOwner = ArbOwner(0x0000000000000000000000000000000000000070);
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: why is this address used?

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+import "forge-std/Script.sol";
+
+import "@arbitrum/nitro-contracts/src/precompiles/ArbOwner.sol";
+
+/// @notice should be included in an operation batch in the L1 timelock along with SetWasmModuleRootAction
+contract UpgradeArbOSVersionAtTimestampAction {
+    uint64 public immutable newArbOSVersion;
+    uint64 public immutable upgradeTimestamp;
+
+    constructor(uint64 _newArbOSVersion, uint64 _upgradeTimestamp) {
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: do we need uint64 for the arbOSVersion?

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        newArbOSVersion = _newArbOSVersion;
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: consider adding error checking to ensure that the values passed here are valid

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol [hunk 12]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: Apache-2.0
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: Not sure which license we'd want to use but we normally default to UNLICENSED

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol [hunk 13]
```diff
@@ -0,0 +1,107 @@
+pragma solidity ^0.8.9;
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: add the SPDX license identifier

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+            TransparentUpgradeableProxy(payable(address(IRollupCore(rollup).challengeManager())));
+        address chalManImpl = ProxyAdmin(proxyAdmin).getProxyImplementation(challengeManager);
+        ProxyAdmin(proxyAdmin).upgradeAndCall(
+            challengeManager,
+            chalManImpl, // Use the rollups current challenge manager as we only need to upgrade the OSP
+            abi.encodeWithSelector(IChallengeManager.postUpgradeInit.selector, IOneStepProofEntry(newOspEntry), currentWasmModuleRoot, IOneStepProofEntry(currentOspEntry))
+        );
+        address postUpgradeChalManAddr = ProxyAdmin(proxyAdmin).getProxyImplementation(challengeManager);
+
+        if(postUpgradeChalManAddr != chalManImpl){
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: i like this check!

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+    constructor(
+        address _newOspEntry,
+        bytes32 _newWasmModuleRoot,
+        address _currentOspEntry,
+        bytes32 _currentWasmModuleRoot,
+        address _rollup,
+        address _proxyAdmin
+    ){
+        newOspEntry = _newOspEntry;
```
Comment (alysiahuggins) [inline contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: perhaps add some error checking for these variables in the constructor

File: contracts/tests/EspressoMigrationAction.t.sol [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+        vm.setEnv("PROXY_ADMIN", Strings.toHexString(uint256(uint160(computeCreateAddress(address(rollupCreator), 1)))));
+        vm.setEnv("NEW_WASM_MODULE_ROOT", Strings.toHexString(uint256(keccak256("newRoot"))));
+        vm.setEnv("CURRENT_WASM_MODULE_ROOT", Strings.toHexString(uint256(keccak256("wasm"))));
+        bytes memory data = abi.encodeWithSelector(
+            OspMigrationAction.perform.selector  
+        );
+        
+        address migration = address(new EspressoOspMigrationAction());
+
+        vm.prank(rollupOwner);
```
Comment (alysiahuggins) [inline contracts/tests/EspressoMigrationAction.t.sol:218]: consider adding more tests that check for negative scenarios e.g. expecting a revert of a non-rollup owner tries to execute the upgrade

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 9 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: This magic value is just the location of the [ArbOwner precompile](https://github.com/EspressoSystems/nitro-espresso-integration/blob/integration/precompiles/ArbOwner.go) that lives on all arbitrum orbit chains at a [specific address](https://docs.arbitrum.io/build-decentralized-apps/precompiles/reference).

I'll leave a comment explaining it in the code.

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 12 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: Ill go ahead and add UNLICENSED to all of the files that require it.

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 8 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: This will be called a decent amount of time before the network upgrade is supposed to take place. This will schedule a particular function inside the nodes of an orbit network to run at the given Unix timestamp.

The hardcoded 35 comes from a discussion I had with Mathis when we were first pursuing designating this as an ArbOS upgrade.

Basically, the way ArbOS upgrades are [performed](https://github.com/EspressoSystems/nitro-espresso-integration/blob/2ac2c23db7e66f6a92b3e9b3163d972389d7ea10/arbos/arbosState/arbosstate.go#L253) is via a switch statement that has some pre-defined gaps in version numbers left open for orbit chains. We chose 35 as it leaves some room for previous upgrades, as well as room for future upgrades in the most recent gap.

File: contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol (hunk 15 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/EspressoOspMigrationAction.sol]: If we add the checks in the perform function to the constructor we can, in theory, remove them from the perform function to avoid using extra gas correct?

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 10 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: I would argue that most of the way this is done doesn't "need" to happen this way, but I wrote these contracts to broadly conform with the guidance in the arbitrum governance repo [here](https://github.com/ArbitrumFoundation/governance/blob/main/src/gov-action-contracts/README.md)

For the sake of time we aren't using address registries and are instead using env variables for the generic and proposal specific version of the contracts. I'm just doing it this way to minimize the amount of future work to convert this to a reasonable proposal for the Arbitrum DAO

File: contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example (hunk 5 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: This might just be from my lack of familiarity with forge and foundry. I couldn't get the env vars to be visible to the tests unless I exported them. [Sourcing the file without exporting them](https://github.com/foundry-rs/foundry/issues/2430#issuecomment-1191828125) doesn't propagate the vars to any child processes, e.g. `forge test`.
 If there is a different way I'm supposed to supply forge with the env vars I would love to know about it!

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 1 above)
Comment (zacshowa) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: This is what I originally thought. However, per the [governance action contract guidelines ](https://github.com/ArbitrumFoundation/governance/blob/main/src/gov-action-contracts/README.md) this seems to be the proper format to write these contracts for an eventual proposal to the Arbitrum DAO.
In theory this also allows us to create more specific proposals by writing more wrapper contracts like the EspressoArbOSUpgrade that utilizes the general

File: contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example (hunk 5 above)
Comment (sveitser) [reply contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: If the file is called .env exactly it is special and automatically picked up by foundry and many other tools. Unfortunately it has to be exactly .env in the case of foundry.

Comment (alysiahuggins) [reply contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: you can do `source .env-espresso-osp-migration && forge test <>` to be able to have a foundry script pick up env variables that's not in the expected `.env` file.

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 8 above)
Comment (alysiahuggins) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: ahh interesting - let's add that in the comments then, and maybe leave instructions in the readme if it ever has to change

File: contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol (hunk 10 above)
Comment (alysiahuggins) [reply contracts/parent-chain/contract-upgrades/EspressoArbOSUpgrade.sol]: ok great!

File: contracts/parent-chain/contract-upgrades/DeployEspressoOsp.s.sol [hunk 17]
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import "forge-std/Script.sol";
+import "nitro-contracts.git/osp/OneStepProverMemory.sol";
```
Comment (sveitser) [inline contracts/parent-chain/contract-upgrades/DeployEspressoOsp.s.sol]:
```suggestion
import "nitro-contracts/osp/OneStepProverMemory.sol";
```

Currently does not matche the submodule path.

cc @zacshowa

Comment (sveitser) [inline contracts/parent-chain/contract-upgrades/DeployEspressoOsp.s.sol]:
```suggestion
import "nitro-contracts/osp/OneStepProverMemory.sol";
```

Currently does not match the submodule path. Following lines too.

cc @zacshowa

File: contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example (hunk 5 above)
Comment (sveitser) [reply contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: That's true but then breaks the ability of being able to override with env vars. So this can get messy when used in scripts if one expects environment variable that are already set to take precedence.

Comment (alysiahuggins) [reply contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: that's true. we aren't using the .env file for secrets anymore but only for config (since that's how it's been set up historically). So should this go into the .env file?

Comment (sveitser) [reply contracts/parent-chain/contract-upgrades/example-env/.env-espresso-osp-migration.example:1]: It seems to be how existing orbit upgrades are configured in the orbit-actions repo. I think it's a good idea to stick to how things are done in that repo unless it really gets in our way.

File: migration/README.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+
+    - Parent chain id: The chain id of the parent chain that has your rollup contracts
+
+    - Child chain id: The chain id for your orbit chain
+    
+    - Parent chain RPC URL: The RPC URL for the parent chain that has your rollup contracts 
+    
+    - Child chain RPC URL: The RPC URL for your orbit chain
+
+    - Hotshot address: The address of the hotshot client contract on the parent chain.
```
Comment (sveitser) [inline migration/README.md]:
```suggestion
    - Hotshot address: The address of the hotshot light client contract on the parent chain.
```

File: migration/README.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+
+    - Proxy admin: The address of the proxy admin for your rollup on the parent chain
+
+    - Upgrade timestamp: The unix timestamp you wish to perform the ArbOS upgrade at.
+
+We would reccommend that you aggregate these in a .env file in variables of the same names to the ones in .example-env so that you can export them when running the commands presented in future sections of this guide.
+
+#### A note on Upgrade Executors
+
+This migration presumes there to be upgrade executor contracts that are chain/rollup owners on both the parent chain, as well as the child chain. If you do not have an upgrade executor on the child chain, the contract execution commands at the end of this guide will not work.
```
Comment (sveitser) [inline migration/README.md:51]: 👍

File: migration/README.md [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+## Contract Deployments
+
+### Parent chain deployments
+
+The parent chain is where most of the contract deployments will happen for the migration to espresso. There are two forge scripts that you need to deploy on the parent chain. These scripts are located in the following directories:
+
+```
+orbit-actions/contracts/parent-chain/espresso-migration/DeployEspressoOsp.s.sol
```
Comment (sveitser) [inline migration/README.md]: You can use markdown links relative to files in this repo.

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links

File: migration/README.md [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+
+These must ***MUST*** be deployed in the following order:
+
+1. DeployEspressoOsp.s.sol
+2. DeployEspressoOspMigrationAction.s.sol
+
+In between these deployments you need to record the address of the newly deployed OneStepProverEntry contract and export it in the env var `NEW_OSP_ENTRY` 
+
+**IMPORTANT NOTE:**
+    These commands will depend on certain environment variables being populated. If you havent already make sure to source the env file with all of the pre-requisite information.
```
Comment (sveitser) [inline migration/README.md:81]: `forge` should pick up the values in the `.env` file automatically.

Ah the commands below also need use the variables directly.

File: migration/README.md [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+```
+orbit-action/contracts/parent-chain/espresso-migration/DeployEspressoOspMigrationAction.s.sol
+```
+
+These must ***MUST*** be deployed in the following order:
+
+1. DeployEspressoOsp.s.sol
+2. DeployEspressoOspMigrationAction.s.sol
+
+In between these deployments you need to record the address of the newly deployed OneStepProverEntry contract and export it in the env var `NEW_OSP_ENTRY`
```
Comment (sveitser) [inline migration/README.md:72]: Is this obvious from the output? Or should we add a `jq` command here?

**Review by Sneh1999 — COMMENTED** [review]
nit: @zacshowa  can you run [forge fmt](https://book.getfoundry.sh/reference/cli/forge/fmt?highlight=forge%20fmt#forge-fmt) to format the contracts?

File: contracts/parent-chain/espresso-migration/EspressoOspMigrationAction.sol [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+            abi.encodeWithSelector(IChallengeManager.postUpgradeInit.selector, IOneStepProofEntry(newOspEntry), currentWasmModuleRoot, IOneStepProofEntry(currentOspEntry))
+        );
+        address postUpgradeChalManAddr = ProxyAdmin(proxyAdmin).getProxyImplementation(challengeManager);
+
+        if(postUpgradeChalManAddr != chalManImpl){
+            revert ChallengeManagerUpdated(postUpgradeChalManAddr);
+        }
+        IOneStepProofEntry newOsp = IChallengeManager(address(challengeManager)).osp();
+
+        if(newOsp != IOneStepProofEntry(newOspEntry) ){
```
Comment (Sneh1999) [inline contracts/parent-chain/espresso-migration/EspressoOspMigrationAction.sol]: Is there a reason why we are storing `newOspEntry` in the constructor? If we store it in the constructor wont we have to deploy a new contract everytime we have to perform an upgrade?

File: migration/README.md (hunk 22 above)
Comment (zacshowa) [reply migration/README.md:72]: I think the output of that command would be obvious if you're running the same verbosity level on the forge script. But I think it also makes sense to add the command anyways. as we know exactly how forge will output the broadcast file.

File: contracts/parent-chain/espresso-migration/EspressoOspMigrationAction.sol (hunk 23 above)
Comment (zacshowa) [reply contracts/parent-chain/espresso-migration/EspressoOspMigrationAction.sol]: I do see your point that we would have to deploy a new version of the contract for each migration. I think that would be more in line with the approach intended for Arbitrum upgrades.

Typically these contracts are supposed to be instantiated with specific magic values for the constructor arguments that make it easier for the Arbitrum DAO to see exactly what would happen when the upgrade is ran. 

I might be wrong but wouldn't each rollup want a distinct OSP entry anyways?

Comment (Sneh1999) [reply contracts/parent-chain/espresso-migration/EspressoOspMigrationAction.sol]: oh okay, if this is the way Arbitrum does it - I think it is okay

**Review by Sneh1999 — APPROVED** [review]
